### PR TITLE
feat: api proxy support - SQSERVICES-1794

### PR DESCRIPTION
**Pull request by netbe**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1794" title="SQSERVICES-1794" target="_blank">SubtaskSQSERVICES-1794</a>  Rework App Network Calls Order to account for Proxy
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

_Briefly describe the issue you have solved or implemented with this pull request. If the PR contains multiple issues, use a bullet list._

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

_Briefly describe the solutions you have implemented for the issues explained above._

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Source/SessionManager/SessionFactories.swift [hunk 1]
```diff
@@ -83,8 +85,19 @@ open class AuthenticatedSessionFactory {
...
+    }
+
+    // MARK: - Private
+
+    private(set) var proxyUsername: String?
+    private(set) var proxyPassword: String?
 }
 
+// MARK: -
+
```
Comment (David-Henner) [inline Source/SessionManager/SessionFactories.swift:100]:
```suggestion
```

File: Source/SessionManager/SessionFactories.swift [hunk 2]
```diff
@@ -27,12 +27,7 @@ open class AuthenticatedSessionFactory {
     var analytics: AnalyticsType?
     let application: ZMApplication
 
-    var environment: BackendEnvironmentProvider {
-        didSet {
-            reachability = environment.reachability
-        }
```
Comment (David-Henner) [inline Source/SessionManager/SessionFactories.swift:33]: why is this removed?

File: Source/SessionManager/SessionManager.swift [hunk 3]
```diff
@@ -458,13 +476,65 @@ public final class SessionManager: NSObject, SessionManagerType {
...
         pushLog.safePublic("registering for voip push token")
         self.pushRegistry.delegate = self
         let pkPushTypeSet: Set<PKPushType> = [PKPushType.voIP]
         self.pushRegistry.desiredPushTypes = pkPushTypeSet
     }
 
+    var proxyCredentials: ProxyCredentials? {
+        didSet {
+
+        }
```
Comment (David-Henner) [inline Source/SessionManager/SessionManager.swift:516]: is this meant to be like this?

File: Source/SessionManager/SessionManager.swift [hunk 4]
```diff
@@ -251,13 +251,12 @@ public final class SessionManager: NSObject, SessionManagerType {
...
         didSet {
+            reachability.tearDown()
+            reachability = environment.reachabilityWrapper()
             authenticatedSessionFactory.environment = environment
-            unauthenticatedSessionFactory.environment = environment
-            reachability = environment.reachability
-
-            // We need a new resolver for the new backend environment.
-            apiVersionResolver = createAPIVersionResolver()
-            resolveAPIVersion()
```
Comment (David-Henner) [inline Source/SessionManager/SessionManager.swift:260]: since this is removed, where do we resolve the api version instead?

File: Source/SessionManager/SessionManager.swift [hunk 5]
```diff
@@ -251,13 +251,12 @@ public final class SessionManager: NSObject, SessionManagerType {
...
+            authenticatedSessionFactory.environment = environment
```
Comment (David-Henner) [inline Source/SessionManager/SessionManager.swift:257]: This line exist twice (line 257 & 256), did you mean to write `unauthenticatedSessionFactory.environment = environment` ?

File: Source/SessionManager/SessionManager.swift [hunk 6]
```diff
@@ -290,6 +289,8 @@ public final class SessionManager: NSObject, SessionManagerType {
 
     var apiVersionResolver: APIVersionResolver?
 
+    private(set) var unauthenticatedTransportSessionReady: Bool = false
```
Comment (David-Henner) [inline Source/SessionManager/SessionManager.swift:292]:
```suggestion
    private(set) var isUnauthenticatedTransportSessionReady: Bool = false
```

File: Source/SessionManager/SessionManager.swift [hunk 7]
```diff
@@ -344,51 +355,56 @@ public final class SessionManager: NSObject, SessionManagerType {
...
-                if blacklisted {
-                    self.isAppVersionBlacklisted = true
-                    self.delegate?.sessionManagerDidBlacklistCurrentVersion(reason: .appVersionBlacklisted)
-                    // When the application version is blacklisted we don't want have a
-                    // transition to any other state in the UI, so we won't inform it
-                    // anymore by setting the delegate to nil.
-                    self.delegate = nil
+       configureBlacklistDownload()
+
+        self.memoryWarningObserver = NotificationCenter.default.addObserver(
```
Comment (David-Henner) [inline Source/SessionManager/SessionManager.swift:364]: the indentation seems off here

File: Source/UserSession/ZMBlacklistDownloader.h [hunk 8]
```diff
@@ -28,6 +28,9 @@
 /// Creates a downloader that will download the blacklist file at regular intervals and invokes the completion handler on the main queue when a blacklist is available
 - (instancetype)initWithDownloadInterval:(NSTimeInterval)downloadInterval
                              environment:(id<BackendEnvironmentProvider>)environment
+                           proxyUsername:(NSString *)proxyUsername
+                           proxyPassword:(NSString *)proxyPassword
+                                ready:(BOOL)ready
```
Comment (David-Henner) [inline Source/UserSession/ZMBlacklistDownloader.h:33]: indentation looks off

File: Source/SessionManager/BackendEnvironmentProvider+Reachability.swift [hunk 9]
```diff
@@ -18,12 +18,90 @@
...
+            } else if !enabled {
+                safeReachability?.tearDown()
+                safeReachability = nil
+            }
+        }
+    }
+    var reachabilityClosure: () -> WrappedReachability
+    private var safeReachability: WrappedReachability?
+
+    init(enabled: Bool, reachabilityClosure: @escaping () -> WrappedReachability) {
```
[the remark this answers is not in the document]
Comment (netbe) [reply Source/SessionManager/BackendEnvironmentProvider+Reachability.swift:95]: the idea is to delay the creation of Reachability. At initialization Reachability sends network activity. We want to delay this  until we're ready to initialize network (proxy received or user input that there is no proxy settings)

File: Source/SessionManager/BackendEnvironmentProvider+Reachability.swift [hunk 10]
```diff
@@ -18,12 +18,90 @@
...
+        self.enabled = enabled
+        self.reachabilityClosure = reachabilityClosure
+        super.init()
+        if enabled {
+            self.safeReachability = reachabilityClosure()
+        }
```
[the remark this answers is not in the document]
Comment (netbe) [reply Source/SessionManager/BackendEnvironmentProvider+Reachability.swift:101]: can't since we don't want a Reachability. Though with the closure maybe if we deal properly with tearDown...

File: Source/SessionManager/SessionFactories.swift (hunk 2 above)
Comment (netbe) [reply Source/SessionManager/SessionFactories.swift:33]: I don't remember, maybe not to retain again reachability once it's replaced

File: Source/SessionManager/SessionManager.swift (hunk 5 above)
Comment (netbe) [reply Source/SessionManager/SessionManager.swift:257]: yes

File: Source/SessionManager/SessionManager.swift (hunk 4 above)
Comment (netbe) [reply Source/SessionManager/SessionManager.swift:260]: we resolve it at login

File: Source/SessionManager/SessionManager.swift (hunk 3 above)
Comment (netbe) [reply Source/SessionManager/SessionManager.swift:516]: to be removed

File: Source/SessionManager/APIVersionResolver.swift [hunk 11]
```diff
@@ -48,16 +48,22 @@ final class APIVersionResolver {
 
     // MARK: - Methods
 
-    func resolveAPIVersion(completion: @escaping () -> Void = {}) {
+    func resolveAPIVersion(completion: @escaping (Error?) -> Void = { _ in }) {
+        // TODO: check if it's been 24hours and proceed or not
```
[the remark this answers is not in the document]
Comment (netbe) [reply Source/SessionManager/APIVersionResolver.swift:52]: yes

File: Source/SessionManager/BackendEnvironmentProvider+Reachability.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (netbe) [reply Source/SessionManager/BackendEnvironmentProvider+Reachability.swift:101]: we don't want to recreate a reachability each time though only when changing enabled

File: Source/SessionManager/SessionFactories.swift (hunk 1 above)
Comment (netbe) [reply Source/SessionManager/SessionFactories.swift:100]: I think the separator was here before to separate between UnauthenticatedSessionFactory and AuthenticatedSessionFactory

File: Source/SessionManager/SessionManager.swift [hunk 12]
```diff
@@ -664,7 +751,9 @@ public final class SessionManager: NSObject, SessionManagerType {
 
             return
         }
-
+        // TODO: is it the right place to activate when user is logged in
+        // we can go and activate Reachability
+        markSessionsAsReady(true)
```
Comment (johnxnguyen) [inline Source/SessionManager/SessionManager.swift:756]: I don't quite understand this TODO. Is it still relevant?

Comment (netbe) [reply Source/SessionManager/SessionManager.swift:756]: I'll give it a try

Comment (netbe) [reply Source/SessionManager/SessionManager.swift]: works within tests and no change in the app 👍

**Review by johnxnguyen — APPROVED** [review]

**Comment by netbe** [comment]
release major

**Comment by zenkins** [comment]
Hello, @netbe! Merging PR for command **release major** for PR #**1800**. Will create a **major** release after merged.
Next version is 436.0.0.
